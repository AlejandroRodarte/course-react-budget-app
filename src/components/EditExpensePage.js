import React from 'react';

import { connect } from 'react-redux';

import { startEditExpense, startRemoveExpense } from '../actions/expenses';

import ExpenseForm from './ExpenseForm';

export class EditExpensePage extends React.Component {

    onSubmit = expense => {
        this.props.startEditExpense(this.props.match.params.id, expense);
        this.props.history.push('/');
    };

    onClick = () => {
        this.props.startRemoveExpense(this.props.match.params.id)
        this.props.history.push('/');
    };

    render() {
        return (
            <div>

                <ExpenseForm
                    onSubmit={ this.onSubmit }
                    expense={ this.props.expense }
                />

                <button
                    onClick={ this.onClick }
                >
                    Remove
                </button>

            </div>
        );
    }

}

const mapStateToProps = (state, props) => ({
    expense: state.expenses.find(expense => expense.id === props.match.params.id)
});

const mapDispatchToProps = dispatch => ({
    startEditExpense: (id, updates) => dispatch(startEditExpense({ id, updates })),
    startRemoveExpense: id => dispatch(startRemoveExpense(id))
});

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);