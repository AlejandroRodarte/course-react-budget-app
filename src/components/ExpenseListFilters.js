import React from 'react';

import { connect } from 'react-redux';

import { setTextFilter, sortByAmount, sortByDate } from '../actions/filters';

const getSortByAction = option => {
    if (option === 'date') {
        return sortByDate;
    } else if (option === 'amount') {
        return sortByAmount;
    } else {
        return sortByDate;
    }
};

const ExpenseListFilters = ({ filters, dispatch }) => (

    <div>

        <input 
            type="text" 
            value={ filters.text }
            onChange={ e => dispatch(setTextFilter(e.target.value)) }
        />

        <select
            value={ filters.sortBy }
            onChange={ e => dispatch(getSortByAction(e.target.value)()) }
        >

            <option value="date">
                Date
            </option>
            
            <option value="amount">
                Amount
            </option>

        </select>

    </div>

);

const mapStateToProps = state => ({
    filters: state.filters
});

export default connect(mapStateToProps)(ExpenseListFilters);