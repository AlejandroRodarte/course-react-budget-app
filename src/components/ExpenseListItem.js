import React from 'react';

import moment from 'moment';

import numeral from 'numeral';

import { Link } from 'react-router-dom';

const ExpenseListItem = ({ id, description, amount, createdAt }) => (

    <Link
        class="list-item"
        to={ `/edit/${id}` }
    >

        <div>
        
            <h3>
                { description }
            </h3>
        
            <span>
                { moment(createdAt).format('MMMM Do, YYYY') }
            </span>

        </div>

        <h3>
            { numeral(amount / 100).format('$0,0.00') }
        </h3>

    </Link>

);

export default ExpenseListItem;