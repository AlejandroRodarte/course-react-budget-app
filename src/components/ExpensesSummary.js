import React from 'react';

import { Link } from 'react-router-dom';

import { connect } from 'react-redux';

import numeral from 'numeral';

import getVisibleExpenses from '../selectors/expenses';
import selectExpensesTotal from '../selectors/expenses-total';

export const ExpensesSummary = ({ expenseCount, expensesTotal }) => (

    <div
        className="page-header"
    >

        <div
            className="content-container"
        >

            <h1
                className="page-header__title"
            >

                Viewing &nbsp;

                <span>
                    { expenseCount }
                </span>

                &nbsp; expense{ expenseCount !== 1 && 's' } totalling &nbsp;
                
                <span>
                    { numeral(expensesTotal / 100).format('$0,0.00') }
                </span>

            </h1>

            <div
                className="page-header__actions"
            >
            
                <Link
                    className="button"
                    to="/create"
                >
                    Add Expense
                </Link>

            </div>

        </div>

    </div>

);

const mapStateToProps = state => {

    const visibleExpenses = getVisibleExpenses(state.expenses, state.filters);

    return {
        expenseCount: visibleExpenses.length,
        expensesTotal: selectExpensesTotal(visibleExpenses)
    };

};

export default connect(mapStateToProps)(ExpensesSummary);
