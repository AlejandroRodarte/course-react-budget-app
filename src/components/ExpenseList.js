import React from 'react';

import { connect } from 'react-redux';

import ExpenseListItem from './ExpenseListItem';

import getVisibleExpenses from '../selectors/expenses';

export const ExpenseList = props => (

    <div
        className="content-container"
    >

        <div
            className="list-header"
        >

            <div
                class="show-for-mobile"
            >
                Expenses
            </div>
        
            <div
                class="show-for-desktop"
            >
                Expense
            </div>
        
            <div
                class="show-for-desktop"
            >
                Amount
            </div>

        </div>

        {
            props.expenses.length === 0 ? (
                <p>No expenses</p>
            ): (
                props.expenses.map(expense => (
                    <ExpenseListItem key={ expense.id } { ...expense } />
                ))
            )
        }

    </div>

);

const mapStateToProps = state => ({
    expenses: getVisibleExpenses(state.expenses, state.filters)
});

export default connect(mapStateToProps)(ExpenseList);
