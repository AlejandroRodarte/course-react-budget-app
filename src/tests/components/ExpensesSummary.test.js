import React from 'react';

import { shallow } from 'enzyme';

import { ExpensesSummary } from '../../components/ExpensesSummary';

test('Should render ExpensesSummary with a single expense', () => {
    
    const wrapper = shallow(
        <ExpensesSummary
            expenseCount={ 2 }
            expensesTotal={ 9434 }
        />
    );

    expect(wrapper).toMatchSnapshot();

});

test('Should render ExpensesSummary with a multiple expenses', () => {

    const wrapper = shallow(
        <ExpensesSummary
            expenseCount={ 1 }
            expensesTotal={ 9434 }
        />
    );

    expect(wrapper).toMatchSnapshot();

});
