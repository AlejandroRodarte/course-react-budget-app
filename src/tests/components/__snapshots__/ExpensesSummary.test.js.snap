// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render ExpensesSummary with a multiple expenses 1`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <h1
      className="page-header__title"
    >
      Viewing  
      <span>
        1
      </span>
        expense
       totalling  
      <span>
        $94.34
      </span>
    </h1>
    <div
      className="page-header__actions"
    >
      <Link
        className="button"
        to="/create"
      >
        Add Expense
      </Link>
    </div>
  </div>
</div>
`;

exports[`Should render ExpensesSummary with a single expense 1`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <h1
      className="page-header__title"
    >
      Viewing  
      <span>
        2
      </span>
        expense
      s
       totalling  
      <span>
        $94.34
      </span>
    </h1>
    <div
      className="page-header__actions"
    >
      <Link
        className="button"
        to="/create"
      >
        Add Expense
      </Link>
    </div>
  </div>
</div>
`;
